server/server.go

```diff
@@ -1656,6 +1665,145 @@ func (s *StanServer) start(runningState State) error {
 	return nil
 }
 
+// startChannelGossiping sets up a NATS subscription to a cluster gossip
+// subject to receive messages from the cluster on what channels exist. It also
+// starts a goroutine which periodically publishes the channels this server
+// knows about. When a server discovers it's missing channels, it requests the
+// missing channels from another server and creates them. This should only be
+// called if the server is running in clustered mode.
+func (s *StanServer) startChannelGossiping() error {
+	// Setup sub for receiving channel lists for reconciliation.
+	if _, err := s.ncr.Subscribe(s.getChannelReconcileInbox(), s.reconcileChannels); err != nil {
+		return err
+	}
+
+	// Setup sub for receiving channel gossip.
+	gossipInbox := fmt.Sprintf("%s.%s.channels", defaultGossipPrefix, s.opts.ID)
+	if _, err := s.ncr.Subscribe(gossipInbox, s.processChannelGossip); err != nil {
+		return err
+	}
+
+	// Setup sub for receiving channel requests.
+	reqInbox := fmt.Sprintf("%s.%s.channels.request.%s", defaultGossipPrefix, s.opts.ID, s.opts.Clustering.NodeID)
+	if _, err := s.ncr.Subscribe(reqInbox, s.processChannelListRequest); err != nil {
+		return err
+	}
+
+	// Start publishing channel info.
+	go func() {
+		rand.Seed(time.Now().UnixNano())
+		for {
+			// Publish on a fixed interval with some random jitter.
+			randomDelay := time.Duration(rand.Intn(6)) * time.Second
+			select {
+			case <-time.After(s.opts.Clustering.GossipInterval + randomDelay):
+				s.channels.RLock()
+				count := uint32(len(s.channels.channels))
+				s.channels.RUnlock()
+				data, err := (&spb.ChannelGossip{NodeID: s.opts.Clustering.NodeID, NumChannels: count}).Marshal()
+				if err != nil {
+					panic(err)
+				}
+				s.ncr.PublishRequest(gossipInbox, reqInbox, data)
+			case <-s.shutdownCh:
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// reconcileChannels is a NATS handler that handles asynchronous responses from
+// a peer after requesting the list of channels. This is done asynchronously
+// since the list can exceed the max NATS message size, so we stream the list
+// in chunks asynchronously. Reconciling channels should be idempotent.
+func (s *StanServer) reconcileChannels(m *nats.Msg) {
+	// Message cannot be empty, we are supposed to receive
+	// a spb.CtrlMsg_Partitioning protocol.
+	if len(m.Data) == 0 {
+		return
+	}
+	req := spb.CtrlMsg{}
+	if err := req.Unmarshal(m.Data); err != nil {
+		s.log.Errorf("Error processing channel reconcile request: %v", err)
+		return
+	}
+	// Ignore messages from ourselves.
+	if req.ServerID == s.serverID {
```

Comment:
Looks like this could not happen, right? A server is receiving the list on an private subject and it would not have send the request to itself.

Reply:
In partitioning code, the subject was public, so a server would have to exclude control messages that it received from itself.

Reply:
Yes, I will remove the check as it's unnecessary.